rfcs/20200827-keras-cv-scoping-design.md

```diff
@@ -0,0 +1,115 @@
+# Keras CV
+
+| Status        | Proposed      |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Zhenyu Tan (tanzheny@google.com), Francois Chollet (fchollet@google.com) |
+| **Updated**   | 2020-08-27                                           |
+
+
+## Objective
+
+This document describes the scope of the [keras-cv](https://github.com/keras-team/keras-cv) package, especially:
+- What use cases `keras-cv` should cover
+- Boundaries between `keras-cv` and [TensorFlow Addons](https://github.com/tensorflow/addons)
+- Boundaries between `keras-cv` and [TensorFlow model garden](https://github.com/tensorflow/models)
+- Boundaries between `keras-cv` and [tf.keras.applications](https://keras.io/api/applications/)
+
+## Motivation
+
+Computer vision (CV) is a major application area for our users.
+Keras on its own provides good support for image classification tasks, in particular via `tf.keras.applications`.
+However, a Keras-native modeling solutions for more advanced tasks,
+such as object detection, instance segmentation, etc., is still lacking.
+
+As a result, the open-source community has rolled out many different solutions for these use cases,
+made available via PyPI and GitHub. These third-party solutions are not always kept up to date, and
+many still rely on the legacy multi-backend Keras. They also raise the issue of API standardization.
+
+To fix this, we want machine learning engineers to have access to a standard Keras-native,
+optimized, and well-tested set of components to build their advanced computer vision models.
+
+This provides key user benefits:
+
+- The package would be first-party and thus always up to date with modern best practices.
+- High code quality and testing standards and strict quality control: same level of trust as core Keras
+- A shared API standard across the community
+- Ability for the open-source community to build more advanced solutions *on top* of this package instead of reinventing it
+
+## Design Proposal
+
+`keras-cv` will provide components that cover the following areas:
+
+- Object Detection tasks.
+- Instance Segmentation tasks.
+- Semantic Segmentation tasks.
+- Keypoint Detection tasks.
+- Video Classification tasks.
+- Object Tracking tasks.
+
+Specifically, for Object Detection tasks, `keras-cv` will include most anchor-based modules:
+
+- Common objects such as anchor generator, box matcher.
+- Keras layer components such as ROI generator, NMS postprocessor.
```

Comment:
Similar question about custom op kernels as in the NLP RFC.

Reply:
NMS is mostly supported (through 5 different versions). I agree there maybe new custom-ops needed, I believe one thing we haven't mentioned that @bhack asked is what things should go to keras-cv and what goes to tf core. In this specific case, graduating tfa custom ops to tf core, not keras, seems a better option IMO.

Reply:
Mai question probably alteady done in another ticket: are these repository going to host c++ code?
Cause "Keras world" was historically python only.
More in genetsl see also our old thread about custom ops at https://github.com/tensorflow/addons/issues/1752#issuecomment-621322109

Reply:
I am asking this also by an inference point of view. As we could see from the Google mediapipe experience about putting end2end TF model "in production" on different platforms (e.g. also pm devices where the python interpreter it is not available) it still require many c++ calculators with impl sometime depending on external c++ library like OpenCV etc...
Just to make an example see c++ calculators code in the image folder: https://github.com/google/mediapipe/tree/master/mediapipe/calculators/image

Reply:
I think the philosophy here is, we cannot guarantee everything has a tf op. This is not aligned with our future endeavor such as MLIR as well. Instead we should allow compiler to interpret them and break down into simpler ops.
But this is a good point, specifically regarding OpenCV, at least they can be executed wrapped in tf.numpy_function. For training this is good enough (and part of data preprocessing, so if you don't have accelerate support, it's fine). For serving, I can imagine people have their own solutions for optimization

Reply:
This was exactly the point. So why not involve a larger audience on defining a CV dialect for MLIR. This could also cover not differentiable high level operation for transformation/lowering. 

I am not in the compiler domain but some month ago we discussed something like this with @stellaraccident in https://github.com/google/mlir-npcomp/issues/1

Reply:
Sounds good. Meanwhile will this discussion alternate the proposal?

Reply:
It Is hard to say cause I am not in the position to allocate TF teams  resources in a specific direction. :wink: But It would be nice if you will find, as a team, an internal consesus and resources to bootstrap an MVP on this (if It really makes any sense to create a CV dialect)

Reply:
Also as you know by some concrete example we hard on differenti tickets coordination It Is hard in the Wild with a classical approach. 
E.g. Just to mention something old and fresh at the same time https://github.com/tensorflow/addons/pull/914#issuecomment-671643050

Reply:
Specifically related to `this package being pure python`? (keras_cv is on top of Keras IMO)

Reply:
Related to your mentioned `NMS postprocessor` and iter future candidate operators like this one. Being python only doesn't solve the runtime topic about trasformation and compilers. 
Where keras-cv could run? Can run only on targets with python interpreter. 
IMHO It Is a Little bit like when we used PIL preprocessing in Keras instead of rely on TF ops like in the new preprocessing. It Is not really the case of NMS cause we are now at v3 impl in Tensorflow. I picked this just cause you mentioned as an example.